pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,267 @@
+	}
+
+	return nil
+}
+
+func createFirewallRule(ctx context.Context, bastion *extensionsv1alpha1.Bastion, gcpclient gcpclient.Interface, opt *Options) error {
+	rb := &compute.Firewall{
+		Allowed:      []*compute.FirewallAllowed{{IPProtocol: "tcp", Ports: []string{strconv.Itoa(SSHPort)}}},
+		Description:  "Allowed all traffic",
+		Direction:    "INGRESS",
+		TargetTags:   []string{"gardenctl"},
+		Name:         opt.FirewallName,
+		Network:      "projects/" + opt.ProjectID + "/global/networks/" + opt.Shoot.Name,
+		SourceRanges: []string{opt.PublicIP},
+	}
+	_, err := gcpclient.Firewalls().Insert(opt.ProjectID, rb).Context(ctx).Do()
+	if err != nil {
+		return errors.Wrap(err, "could not create firewall rule")
+	}
+
+	logger.Info("Firewall created", "firewall", opt.FirewallName)
+	return nil
+}
+
+func ensureBastionInstance(ctx context.Context, logger logr.Logger, gcpclient gcpclient.Interface, opt *Options) (*bastionEndpoints, error) {
+	// check if the instance already exists and has an IP
+	endpoints, err := getInstanceEndpoints(ctx, gcpclient, opt)
+	if err != nil { // could not check for instance
+		return nil, errors.Wrap(err, "failed to check for GCP Bastion instance")
+	}
+
+	// instance exists, though it may not be ready yet
+	if endpoints != nil {
+		return endpoints, nil
+	}
+
+	logger.Info("Running new bastion instance")
+
+	disk := &compute.Disk{
+		Description: "Gardenctl Bastion disk",
+		Name:        opt.DiskName,
+		SizeGb:      10,
+		SourceImage: "projects/debian-cloud/global/images/family/debian-10",
+		Zone:        opt.Zone,
+	}
+
+	_, err = gcpclient.Disks().Insert(opt.ProjectID, opt.Zone, disk).Context(ctx).Do()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create disk")
+	}
+
+	logger.Info("Disk created", "disk", opt.DiskName)
+
+	attachedDisk := &compute.AttachedDisk{
+		AutoDelete: true,
+		Boot:       true,
+		DiskSizeGb: 10,
+		Source:     fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/disks/%s", opt.ProjectID, opt.Zone, opt.DiskName),
+		Mode:       "READ_WRITE",
+	}
+
+	arr := make([]*compute.AttachedDisk, 1)
+	arr = append(arr, attachedDisk)
+
+	network := &compute.NetworkInterface{
+		Network:       "projects/" + opt.ProjectID + "/global/networks/" + opt.Shoot.Name,
+		Subnetwork:    "regions/" + opt.Shoot.Spec.Region + "/subnetworks/" + opt.Subnetwork,
+		AccessConfigs: []*compute.AccessConfig{{Name: "External NAT", Type: "ONE_TO_ONE_NAT"}},
+	}
+	networkArr := make([]*compute.NetworkInterface, 1)
+	networkArr = append(networkArr, network)
+	machineType := fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/machineTypes/n1-standard-1", opt.ProjectID, opt.Zone)
+
+	insta, err := gcpclient.Instances().Get(opt.ProjectID, opt.Zone, opt.BastionInstanceName).Context(ctx).Do()
+	if err != nil {
+		gerr := err.(*googleapi.Error)
+		if gerr.Code != 404 {
+			return nil, errors.Wrap(err, "failed to get compute instance")
+		}
```

Comment:
committed changes in 7a3202b

Reply:
Please handle it similar like this https://github.com/gardener/gardener-extension-provider-gcp/blob/master/pkg/gcp/client/storage.go#L74-L77
if it is not a googleapi.Error you just return the error. With your change, the original error would be lost.

Reply:
committed in bf5322de87086206f0fce8ae97b98b7602510c5d

Reply:
```go
func getDisk(ctx context.Context, gcpclient gcpclient.Interface, opt *Options) (*compute.Disk, error) {
	disk, err := gcpclient.Disks().Get(opt.ProjectID, opt.Zone, opt.DiskName).Context(ctx).Do()
	if err != nil {
		if googleError, ok := err.(*googleapi.Error); ok && googleError.Code == 404 {
			logger.Info("Disk not found,", "disk_name", opt.DiskName)
			return nil, nil
		} else if googleError, ok := err.(*googleapi.Error); ok && googleError.Code == 409 {
			logger.Info(disk.Name + ":" + "Disk already exists")
			return disk, nil
		}
		return nil, fmt.Errorf("failed to get Disk: %w", err)
	}
	return disk, nil
}
```
@tedteng now you are handling `409` error codes on `get` requests which is rather unusual that such an error code is returned in this case.
Maybe you thought to take it over from the link I sent you above? However the difference is that the [code snipped](https://github.com/gardener/gardener-extension-provider-gcp/blob/master/pkg/gcp/client/storage.go#L74-L77) was for a `create` request instead of a `get`. With the [link](https://github.com/gardener/gardener-extension-provider-gcp/blob/master/pkg/gcp/client/storage.go#L74-L77) I was referring to the handling of the google error and not the error code itself.

Anyhow, please remove the handling for `409` error codes on all the `get` requests again.

Reply:
committed in 9fd710bc730e1dc874e8e32491c5bdb2db9afae0